product_multi_image/models/product_image.py

```diff
@@ -0,0 +1,168 @@
+    def _get_image(self):
+        self.image = False
+        if self.type == 'file':
+            if os.path.exists(self.path):
+                try:
+                    with open(self.path, 'rb') as f:
+                        self.image = base64.b64encode(f.read())
+                except Exception, e:
+                    _logger.error("Can not open the image %s, error : %s",
+                                  self.path, e, exc_info=True)
+            else:
+                _logger.error("The image %s doesn't exist ", self.path)
+        elif self.type == 'url':
+            (filename, header) = urllib.urlretrieve(self.url)
+            try:
+                with open(filename, 'rb') as f:
+                    self.image = base64.b64encode(f.read())
+            except:
+                _logger.error("URL %s cannot be fetched", self.url)
+        else:
+            self.image = self.file_db_store
+
+    @api.multi
+    def _check_filestore(self):
+        """check if the filestore is created, and do it otherwise.
+        """
+        for product_image in self:
+            dir_path = os.path.dirname(product_image.path)
+            try:
+                if not os.path.exists(dir_path):
+                    os.makedirs(dir_path)
+            except OSError, e:
+                raise exceptions.Warning(
+                    _('The image filestore can not be created, %s') % e)
+        return True
+
+    @api.one
+    def _set_image(self):
+        if self.type == 'file':
+            self._check_filestore()
+            with open(self.path, 'w') as ofile:
+                ofile.write(base64.b64decode(self.image))
+        elif self.type == 'url':
+            self.type = 'db'
+            self.file_db_store = self.file
+        else:
+            self.file_db_store = self.file
+
+    type = fields.Selection(
+        selection=[('url', 'URL'), ('file', 'OS file'), ('db', 'DB')])
+    name = fields.Char(string='Image title', required=True)
+    extension = fields.Char(string='file extension', oldname='extention',
+                            readonly=True)
+    file_db_store = fields.Binary(string='Image stored in database')
+    path = fields.Char(
+        string="Image path", compute="_get_path",
+        help="Image path if the image is not a link")
+    url = fields.Char(string='Image remote URL')
+    image = fields.Binary(
+        compute="_get_image", inverse="_set_image", string="File",
+        filters='*.png,*.jpg,*.gif')
+    comments = fields.Text(string='Comments')
+    product_id = fields.Many2one(
+        comodel_name='product.product', string='Product', required=True,
+        ondelete='cascade')
+
+    @api.one
+    def unlink(self):
+        if self.path:
+            os.path.isfile(self.path) and os.remove(self.path)
+        return super(ProductImage, self).unlink()
+
+    @api.constrains('type', 'url')
+    def _check_url(self):
+        if self.type == 'url' and not self.url:
+            raise exceptions.Warning(
```

Comment:
Of course it works because the raise blows up anyway. But if I write code and want to catch errors in `constrains`, I should expect to catch `ValidationError` errors.